File: Programmers/보석 쇼핑.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>
#include <set>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=0;i<(n);++i)

using namespace std;

set<string> st;
unordered_map<string,int> mp;
int cnt,total,mn=987654321;

vector<int> solution(vector<string> gems) {
    vector<int> res;
    int l=0, r=0;
    
    FOR(i,gems.size()) st.insert(gems[i]);
    total=st.size();
    
    
    while(!((r==gems.size()) && cnt<total)) {
        if(cnt==total) {
            if(--mp[gems[l++]]==0) --cnt;
        }
        else {
            if(++mp[gems[r++]]==1) ++cnt;
        }
        
        if(cnt==total) {
            if(r-l<mn) {
                mn=r-l;
                res.clear(); res.push_back(l+1); res.push_back(r);
            }
        }
    }
    
    return res;
}

File: Programmers/Python/보석 쇼핑.py
st = set()
hash = {}

def solution(gems):
    for x in gems:
        st.add(x)
    ln = len(gems)
    sz = len(st)
    l=0
    r=0
    hash[gems[0]] = 1
    res = []
    mn = 987654321
    while r<ln:
        if len(hash) < sz:
            r += 1
            if r == ln:
                break
            if gems[r] in hash:
                num = hash[gems[r]]
                hash[gems[r]] = num + 1
            else:
                hash[gems[r]] = 1
        else:
            if mn > r-l:
                mn = r-l;
                res.clear()
                res.append(l+1)
                res.append(r+1)
            if gems[l] in hash:
                num = hash[gems[l]]
                hash[gems[l]] = num - 1
            if hash[gems[l]] == 0:
                del hash[gems[l]]
            l += 1
    
    return res
